### PR TITLE
MM-25277  Return channels that aren't in a category as part of the Channels/Direct Messages categories

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
When calling the "get category" and "get categories" APIs, any public/private channels that aren't in any category should be returned as part of the Channels category and any DM/GM channels that aren't in any category should be returned as part of the Direct Messages category.

These channels should be sorted alphabetically by display name (even though DMs/GMs don't have a display name set) and they should appear before other channels in the category.
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-25277

### CODE REVIEW

**Comment by hmhealey** [comment]
Looks good to me, but could we maybe add some tests for this?

File: store/storetest/channel_store.go [hunk 1]
```diff
@@ -35,68 +35,68 @@ func cleanupChannels(t *testing.T, ss store.Store) {
...
-	t.Run("MaterializedPublicChannels", func(t *testing.T) { testMaterializedPublicChannels(t, ss, s) })
-	t.Run("GetAllChannelsForExportAfter", func(t *testing.T) { testChannelStoreGetAllChannelsForExportAfter(t, ss) })
-	t.Run("GetChannelMembersForExport", func(t *testing.T) { testChannelStoreGetChannelMembersForExport(t, ss) })
-	t.Run("RemoveAllDeactivatedMembers", func(t *testing.T) { testChannelStoreRemoveAllDeactivatedMembers(t, ss, s) })
-	t.Run("ExportAllDirectChannels", func(t *testing.T) { testChannelStoreExportAllDirectChannels(t, ss, s) })
-	t.Run("ExportAllDirectChannelsExcludePrivateAndPublic", func(t *testing.T) { testChannelStoreExportAllDirectChannelsExcludePrivateAndPublic(t, ss, s) })
-	t.Run("ExportAllDirectChannelsDeletedChannel", func(t *testing.T) { testChannelStoreExportAllDirectChannelsDeletedChannel(t, ss, s) })
-	t.Run("GetChannelsBatchForIndexing", func(t *testing.T) { testChannelStoreGetChannelsBatchForIndexing(t, ss) })
-	t.Run("GroupSyncedChannelCount", func(t *testing.T) { testGroupSyncedChannelCount(t, ss) })
+	// t.Run("Save", func(t *testing.T) { testChannelStoreSave(t, ss) })
```
Comment (hmhealey) [inline store/storetest/channel_store.go:38]: Looks good, but you forgot to uncomment the other test cases.

Comment (reflog) [reply store/storetest/channel_store.go:38]: 🤦

**Comment by reflog** [comment]
@hmhealey - this includes unit tests. see at the bottom (checking the length of Channels in category)

File: store/sqlstore/channel_store.go [hunk 2]
```diff
@@ -3490,6 +3490,38 @@ func (s SqlChannelStore) CreateSidebarCategory(userId, teamId string, newCategor
...
+			sq.NotEq{"ChannelMembers.ChannelId": category.Channels},
+			sq.Eq{"ChannelMembers.UserId": category.UserId},
+			channelTypeFilter,
+			sq.Eq{"Channels.DeleteAt": 0},
+		}).
+		OrderBy("DisplayName DESC").ToSql()
+	if _, err := s.GetReplica().Select(&channels, sql, args...); err != nil {
+		return nil, model.NewAppError("SqlPostStore.GetSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusNotFound)
+	}
+	category.Channels = append(channels, category.Channels...)
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:3521]: Any reason this is not the usual `category.Channels = append(category.Channels, channels...)` ?

File: store/sqlstore/channel_store.go [hunk 3]
```diff
@@ -3490,6 +3490,38 @@ func (s SqlChannelStore) CreateSidebarCategory(userId, teamId string, newCategor
 	}, nil
 }
 
+func (s SqlChannelStore) completePopulatingCategoryChannels(category *model.SidebarCategoryWithChannels) (*model.SidebarCategoryWithChannels, *model.AppError) {
+	if category.Type == model.SidebarCategoryCustom || category.Type == model.SidebarCategoryFavorites {
+		return category, nil
+	}
+
+	var sql string
+	var args []interface{}
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:3499]: Any reason to pre-declare them like this ?

File: store/sqlstore/channel_store.go (hunk 2 above)
Comment (reflog) [reply store/sqlstore/channel_store.go:3521]: @agnivade  yes, this is prepend, not append.

File: store/sqlstore/channel_store.go (hunk 3 above)
Comment (reflog) [reply store/sqlstore/channel_store.go:3499]: no. residue from old code. will cleanup

**Pull request merged**